lib/pmdtester/builders/pmd_report_builder.rb

```diff
@@ -27,21 +27,29 @@ def execute_reset_cmd(type, tag)
         reset_cmd = "git reset --hard #{tag}"
       when 'hg'
         reset_cmd = "hg up #{tag}"
-      else
-        raise Exception, "Unknown #{type} repository"
       end
 
       Cmd.execute(reset_cmd)
     end
 
     def get_projects
-      puts 'Cloning projects started'
+      logger.info 'Cloning projects started'
 
       @projects.each do |project|
         path = project.local_source_path
         clone_cmd = "#{project.type} clone #{project.connection} #{path}"
         if File.exist?(path)
-          puts "Skipping clone, project path #{path} already exists"
+          logger.warn "Skipping clone, project path #{path} already exists"
+          update_cmd = "#{project.type} pull #{project.connection}"
+
+          begin
+            Cmd.execute(update_cmd)
+          rescue CmdException
+            logger.warn "updating repository #{path} failed, remove the repository"
+            FileUtils.remove_dir(path)
```

Comment:
Nice idea, but I'm not sure, if we really should automate this... 

+: It's self healing

-: It might be faster (more efficient), if you manually look into the error and fix it. Depending on the problem (e.g. currently no network connection), we would automatically delete the clone and you end up with nothing. And whoever uses pmdtester should be sure then, to never reuse any local repositories with the tool, since they might get deleted...

So, I'd vote for just exiting the tool in this case of error.